src/devices/src/virtio/net/device.rs

```diff
@@ -299,70 +308,90 @@ impl Net {
         success
     }
 
-    // Copies a single frame from `self.rx_frame_buf` into the guest.
-    fn do_write_frame_to_guest(&mut self) -> std::result::Result<(), FrontendError> {
-        let mut result: std::result::Result<(), FrontendError> = Ok(());
-        // This is safe since we checked in the event handler that the device is activated.
-        let mem = self.device_state.mem().unwrap();
-
-        let queue = &mut self.queues[RX_INDEX];
-        let head_descriptor = queue.pop(mem).ok_or_else(|| {
-            METRICS.net.no_rx_avail_buffer.inc();
-            FrontendError::EmptyQueue
-        })?;
-        let head_index = head_descriptor.index;
-
-        let mut frame_slice = &self.rx_frame_buf[..self.rx_bytes_read];
-        let frame_len = frame_slice.len();
-        let mut maybe_next_descriptor = Some(head_descriptor);
-        while let Some(descriptor) = &maybe_next_descriptor {
-            if frame_slice.is_empty() {
-                break;
+    /// Write a slice in a descriptor chain
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the descriptor chain is too short or
+    /// an inappropriate (read only) descriptor is found in the chain
+    fn write_to_descriptor_chain(
+        mem: &GuestMemoryMmap,
+        data: &[u8],
+        head: DescriptorChain,
+    ) -> std::result::Result<(), FrontendError> {
+        let mut chunk = data;
+        let mut next_descriptor = Some(head);
+
+        while let Some(descriptor) = &next_descriptor {
+            if chunk.is_empty() {
+                METRICS.net.rx_bytes_count.add(data.len());
+                METRICS.net.rx_packets_count.inc();
+                return Ok(());
             }
 
             if !descriptor.is_write_only() {
-                result = Err(FrontendError::ReadOnlyDescriptor);
-                break;
+                return Err(FrontendError::ReadOnlyDescriptor);
             }
 
-            let len = std::cmp::min(frame_slice.len(), descriptor.len as usize);
-            match mem.write_slice(&frame_slice[..len], descriptor.addr) {
+            let len = std::cmp::min(chunk.len(), descriptor.len as usize);
+            // TODO: It is safe to unwrap here. We know that `len` fits in the
+            // descriptor.
```

Comment:
I don't know if this guarantees that we won't receive any error. In the end GuestMemory will issue a `write` syscall. Are we sure that it will complete successfully ? Also are we sure we're not going to receive other errors that are not related to `len` ? It would be risky to do this.

Reply:
I left this a TODO because I wanted to examine it better, however reading from the `write_slice` documentation:

```
Returns an error if ther isn't enuogh space within the container to write the entire slice.
Part of the data may have been copied nevertheless.
```